src/cache.ts

```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+import LRUCache from 'lru-cache';
+import { IHashableDependency, SimpleDependency } from './collector';
+
+type CachedItems = {cachedValues: Array<any>, missedItems: Array<IHashableDependency>};
+
+export class Cache {
+  private cache: LRUCache<string, any>;
+
+  constructor(max: number, maxAge: number) {
+    this.cache = new LRUCache<string, any>({max, maxAge});
+  }
+
+  // returns cachedValues and non cached items.
+  classify(deps: Array<IHashableDependency>): CachedItems {
+    const cachedValues = Array<any>();
+    const missedItems = Array<IHashableDependency>();
+    deps.forEach(d => {
+      const cached = this.cache.get(d.key());
+      if (cached) {
+        cachedValues.push(cached);
+      } else {
+        missedItems.push(d);
+      }
+    });
+    return {cachedValues, missedItems};
```

Comment:
Cached and missed items are of different type. Can we have single type for an item stored in cache? It will add lot of confusion to the reader.

Reply:
@dgpatelgit Cache always store the single type, which is response in this case. Actually the `classify` function of our cache do returns 2 types. For the given list of keys, it returns cached values and missing keys. Cached values will be directly passed to pipeline, missing keys will do REST call to get results.

It is a domain specific cache on top of generic LRU cache, so not a big issue in introducing specialisations.

Reply:
It also has a performance benefit, single pass vs  2 pass.

`n x log n vs n x 2log n`

Reply:
Agree, implementation looks ok with this data.